File: Editorial/8_String_to_Integer_atoi/Go/2_Deterministic_Finite_Automaton.go
type State int

const (
    q0 State = iota
    q1
    q2
    qd
)

type StateMachine struct {
    currentState State
    result, sign int
}

func (sm *StateMachine) toStateQ1(ch byte) {
    if ch == '-' {
        sm.sign = -1
    } else {
        sm.sign = 1
    }
    sm.currentState = q1
}

func (sm *StateMachine) toStateQ2(digit int) {
    sm.currentState = q2
    sm.appendDigit(digit)
}

func (sm *StateMachine) toStateQd() {
    sm.currentState = qd
}

func (sm *StateMachine) appendDigit(digit int) {
    if (sm.result > math.MaxInt32/10) ||
        (sm.result == math.MaxInt32/10 && digit > math.MaxInt32%10) {
        if sm.sign == 1 {
            sm.result = math.MaxInt32
        } else {
            sm.result = math.MinInt32
            sm.sign = 1
        }
        sm.toStateQd()
    } else {
        sm.result = sm.result*10 + digit
    }
}

func (sm *StateMachine) Transition(ch byte) {
    switch sm.currentState {
    case q0:
        if ch == ' ' {
            return
        }
        if ch == '-' || ch == '+' {
            sm.toStateQ1(ch)
        } else if ch >= '0' && ch <= '9' {
            sm.toStateQ2(int(ch - '0'))
        } else {
            sm.toStateQd()
        }
    case q1, q2:
        if ch >= '0' && ch <= '9' {
            sm.toStateQ2(int(ch - '0'))
        } else {
            sm.toStateQd()
        }
    }
}

func (sm *StateMachine) GetInteger() int {
    return sm.sign * sm.result
}

func (sm *StateMachine) getState() State {
    return sm.currentState
}

func myAtoi(s string) int {
    Q := StateMachine{currentState: q0, result: 0, sign: 1}
    for i := 0; i < len(s) && Q.getState() != qd; i++ {
        Q.Transition(s[i])
    }
    return Q.GetInteger()
}

File: Editorial/8_String_to_Integer_atoi/CSharp/2_Deterministic_Finite_Automaton.cs
public enum State { q0, q1, q2, qd }

public class StateMachine {
    private State currentState;
    private long result;
    private int sign;

    public StateMachine() {
        currentState = State.q0;
        result = 0;
        sign = 1;
    }

    private void toStateQ1(char ch) {
        sign = (ch == '-') ? -1 : 1;
        currentState = State.q1;
    }

    private void toStateQ2(int digit) {
        currentState = State.q2;
        appendDigit(digit);
    }

    private void toStateQd() {
        currentState = State.qd;
    }

    public void appendDigit(int digit) {
        if ((result > int.MaxValue / 10) ||
            (result == int.MaxValue / 10 && digit > int.MaxValue % 10)) {
            if (sign == 1) {
                result = int.MaxValue;
            } else {
                result = int.MinValue;
                sign = 1;
            }

            toStateQd();
        } else {
            result = result * 10 + digit;
        }
    }

    public void transition(char ch) {
        if (currentState == State.q0) {
            if (ch == ' ') {
                return;
            } else if (ch == '-' || ch == '+') {
                toStateQ1(ch);
            } else if (char.IsDigit(ch)) {
                toStateQ2(ch - '0');
            } else {
                toStateQd();
            }
        } else if (currentState == State.q1 || currentState == State.q2) {
            if (char.IsDigit(ch)) {
                toStateQ2(ch - '0');
            } else {
                toStateQd();
            }
        }
    }

    public int getInteger() {
        return (int)(sign * result);
    }

    public State getState() {
        return currentState;
    }
}

public class Solution {
    public int MyAtoi(string s) {
        StateMachine Q = new StateMachine();
        for (int i = 0; i < s.Length && Q.getState() != State.qd; ++i) {
            Q.transition(s[i]);
        }

        return Q.getInteger();
    }
}
